underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java

```diff
@@ -802,15 +823,28 @@ public void stopSync(AlluxioURI ufsUri) {
    *
    * @param path file or directory path
    * @param recursive whether to delete path recursively
+   * @param isDirectory whether path is a directory
    * @return true, if succeed
    */
-  private boolean delete(String path, boolean recursive) throws IOException {
+  private boolean delete(String path, boolean recursive, boolean isDirectory) throws IOException {
     IOException te = null;
     FileSystem hdfs = getFs();
     RetryPolicy retryPolicy = new CountingRetry(MAX_TRY);
     while (retryPolicy.attempt()) {
       try {
-        return hdfs.delete(new Path(path), recursive);
+        Path hdfsPath = new Path(path);
+        if (!mTrashEnable) {
+          return hdfs.delete(hdfsPath, recursive);
+        }
+        Optional<Trash> trash = getTrash(hdfs);
+        if (!trash.isPresent()) {
+          return hdfs.delete(hdfsPath, recursive);
+        }
+        // move to trash
+        if (isDirectory && !recursive) {
+          return false;
+        }
+        return trash.get().moveToTrash(hdfsPath);
```

Comment:
Are you suggesting that after a failed deletion in the recycle bin, fallback to the regular deletion logic should be implemented? However, this may change the user's expectations. I believe that if deletion fails, the user should be informed of the failure instead of automatically deleting the file for them.

Reply:
Actually, HDFS recycle bin automatically handles files with the same name for us.

Reply:
An example is that if the moveToTrash method is called on the directory /user/hdfs/.Trash/xxx, it will return false. If occur exception like network exception or permission denied exception, it will throw IOException.